### PR TITLE
Improved onnx support

### PR DESCRIPTION
**Pull request by benoitsteiner**

### CODE REVIEW

File: apps/onnx/onnx_converter.cc [hunk 1]
```diff
@@ -7,6 +7,9 @@
 static Halide::Expr div_up(Halide::Expr num, int denom) {
     return Halide::Internal::simplify((num + denom - 1) / denom);
 }
+static Halide::Expr div_down(Halide::Expr num, int denom) {
```
Comment (steven-johnson) [inline apps/onnx/onnx_converter.cc]: nit: blank line between functions

File: apps/onnx/onnx_converter.cc [hunk 2]
```diff
@@ -1892,102 +1925,206 @@ Node convert_split_node(
...
+                    Halide::Internal::simplify(actual_end - actual_start);
+            } else {
+                src_indices.push_back(tgt_indices[i]);
+            }
         } else {
-            src_indices.push_back(tgt_indices[i]);
+            Halide::Expr start = 0;
+            Halide::Expr end = input.shape[i];
+            Halide::Expr step = 1;
+            // Pick slice boundries or keep default values.
```
Comment (steven-johnson) [inline apps/onnx/onnx_converter.cc]: nit: boundaries

File: apps/onnx/onnx_converter.cc [hunk 3]
```diff
@@ -2682,6 +2831,116 @@ Node convert_shrink_node(
     return result;
 }
 
+Node convert_lrn_node(
+    const onnx::NodeProto &node,
```
Comment (steven-johnson) [inline apps/onnx/onnx_converter.cc]: nit: indentation makes args look like start of fn body

File: apps/onnx/onnx_converter.cc [hunk 4]
```diff
@@ -2682,6 +2831,116 @@ Node convert_shrink_node(
...
+    if (input.type == onnx::TensorProto_DataType_FLOAT ||
+        input.type == onnx::TensorProto_DataType_DOUBLE) {
+        Halide::Expr inf_value = input.type == onnx::TensorProto_DataType_FLOAT
+                                     ? Halide::Expr(std::numeric_limits<float>::infinity())
+                                     : Halide::Expr(std::numeric_limits<double>::infinity());
+
+        Halide::Expr pos_inf = detect_positive && input.rep(Halide::_) == inf_value;
+        Halide::Expr neg_inf =
+            detect_negative && input.rep(Halide::_) == -inf_value;
+        // TODO(ataei): Fix Halide's mapping of const nanf into inf_f32.
```
Comment (steven-johnson) [inline apps/onnx/onnx_converter.cc]: Does this need an issue to be opened to track?

Comment (benoitsteiner) [reply apps/onnx/onnx_converter.cc]: I have filed #4100 to request the ability to check for infinite values to be added to Halide proper. It's a much better place for it. I['m not sure about the actual TODO though: @asaadaldien should comment here.

Comment (asaadaldien) [reply apps/onnx/onnx_converter.cc]: This TODO was investigate why evaluating (Halide::Expr == INFINITY) depends on the type ?! e.g in the following example f1, f2 should have the same values but f2 will have 1s wherever g have nanf.
```
f1(halide_var) = Halide::cast<float>(g(halide_var) == INFINITY); 
f2(halide_var) = Halide::cast<int>(g(halide_var) == INFINITY); 
```

A slightly different repro below : 
```
#include <math.h>
#include <Halide.h>

int main(int argc, char** argv) {
  float data[4] = {NAN, INFINITY, -INFINITY, 1.0f};
  Halide::Buffer<float> buff(&data[0], 4);
  Halide::Func f, g;
  Halide::Var x;
  f(x) = Halide::select(buff(x) == INFINITY, 1.0f, buff(x) == -INFINITY, -1.0f, 0.0f);
  g(x) = Halide::select(buff(x) == INFINITY, 1, buff(x) == -INFINITY, -1, 0);
  Halide::Buffer<float> f_out = f.realize(4);
  printf("f = [%.1f, %.1f, %.1f, %.1f], expected = [0.0, 1.0, -1.0, 0.0] \n", f_out(0), f_out(1), f_out(2), f_out(3));
  Halide::Buffer<int> g_out = g.realize(4);
  printf("g = [%d, %d, %d, %d], expected = [0, 1, -1, 0] \n", g_out(0), g_out(1), g_out(2), g_out(3));
  return 0;
}
```

**Comment by steven-johnson** [comment]
LGTM, tests are green

**Pull request merged**